File: clients/go/conventions/goals/conformance_test.go
package goals_test

import (
	"context"
	"encoding/json"
	"fmt"
	"path/filepath"
	"testing"

	conf "github.com/love-lena/sextant/clients/go/conformance"
	"github.com/love-lena/sextant/clients/go/conventions/goals"
	pconf "github.com/love-lena/sextant/protocol/conformance"
)

// The goals convention's conformance suite. It plugs the REAL goal verbs into the
// TASK-183 seam: register them in a conformance.Registry, then ReplayVectors over
// the language-neutral vectors under protocol/conformance/vectors/goals — no
// runner change. These replace the fixture vector as the first real vectors; a
// TypeScript conv/goals (TASK-175) replays the SAME files to prove it is co-equal.

// vectorsDir is the protocol-owned vector root, relative to this test
// (clients/go/conventions/goals → repo root → protocol/conformance/vectors).
func vectorsDir() string {
	return filepath.Join("..", "..", "..", "..", "protocol", "conformance", "vectors")
}

// fixedNow is the timestamp the recorded setCriterion verb stamps, so the
// captured goal.update is byte-stable across runs (time.Now would churn every
// -update). The live SetCriterion takes the real time; only the recorded verb
// pins it.
const fixedNow = "2026-06-19T00:00:00Z"

// setCriterionVerb adapts goals.SetCriterion to a conformance.Verb: decode the
// vector's input, run the verb against the recorder (conf.Ops's method set is a
// superset of goals.Ops, so the recorder is the verb's Ops unchanged), and stamp
// the fixed timestamp so the transcript is deterministic.
func setCriterionVerb(ctx context.Context, ops conf.Ops, input json.RawMessage) error {
	var in goals.SetCriterionInput
	if err := json.Unmarshal(input, &in); err != nil {
		return fmt.Errorf("goals conformance: decode setCriterion input: %w", err)
	}
	_, err := goals.SetCriterion(ctx, ops, in, fixedNow)
	return err
}

func goalsRegistry() *conf.Registry {
	reg := conf.NewRegistry()
	reg.Register("goals", "setCriterion", setCriterionVerb)
	return reg
}

// seedGoal seeds the recorder with the prior goal artifact a setCriterion vector
// reads before it writes. SetCriterion is a read-then-write verb (get the goal,
// mutate a criterion, update); the seed mirrors the bus state it would find live
// and does not appear in the transcript. The seeded goal has the criterion the
// vector flips in a not-yet-target status (so the verb does real work) at the
// revision the vector's artifact.update CAS's against.
func seedGoal(rec *conf.Recorder, v pconf.OpTranscriptVector) {
	if v.Verb != "setCriterion" {
		return
	}
	var in goals.SetCriterionInput
	_ = json.Unmarshal(v.Input, &in)
	// A goal with the target criterion plus a sibling, so the transcript shows the
	// rewrite preserves siblings and the north-star. The criterion starts
	// not-started, so setting any status is a real transition.
	goal := goals.Goal{
		Northstar: "Ship the goals convention",
		Criteria: []goals.Criterion{
			{ID: in.CriterionID, Text: "the criterion under test", Status: goals.StatusNotStarted, Owner: "sirius"},
			{ID: "other", Text: "a sibling criterion", Status: goals.StatusInProgress},
		},
	}
	record, _ := json.Marshal(goal)
	// Revision 4 is the value the vector's artifact.update expectedRev pins.
	rec.SeedArtifact(goals.ArtifactName(in.GoalID), record, 4)
}

// TestGoalsConformance replays the goal vectors against the real verbs. With
// -update it (re)records the on-disk sample vectors from the verbs, so a
// deliberate verb change is a one-command re-record plus a reviewed diff.
func TestGoalsConformance(t *testing.T) {
	dir := vectorsDir()
	reg := goalsRegistry()

	if conf.Updating() {
		recordGoalVectors(t, dir)
	}
	conf.ReplayVectors(t, dir, reg, conf.WithSeed(seedGoal))
}

// recordGoalVectors writes the sample goal vectors from the registered verbs (the
// -update path). Each vector pins one verb call's transcript; the seed supplies
// the prior goal state the read-then-write verb needs.
func recordGoalVectors(t *testing.T, dir string) {
	t.Helper()
	const description = "goals: set a criterion to waiting-on-you — artifact.get, " +
		"artifact.update (CAS the goal with the criterion's status rewritten), then " +
		"message.publish a goal.update on msg.topic.goals."
	input, err := json.Marshal(goals.SetCriterionInput{
		GoalID:      "v0-5-0",
		CriterionID: "c1",
		Status:      goals.StatusWaitingOnYou,
		Headline:    "needs your sign-off",
		Ref:         "the-brief",
		By:          "01CREW",
	})
	if err != nil {
		t.Fatalf("marshal sample input: %v", err)
	}
	v, err := conf.RecordVector(1, "goals", "setCriterion", description, input, setCriterionVerb,
		func(rec *conf.Recorder) {
			seedGoal(rec, pconf.OpTranscriptVector{Verb: "setCriterion", Input: input})
		})
	if err != nil {
		t.Fatalf("record setCriterion: %v", err)
	}
	path := filepath.Join(dir, "goals", "setCriterion.json")
	if err := conf.WriteVector(path, v); err != nil {
		t.Fatalf("write setCriterion: %v", err)
	}
	t.Logf("re-recorded %s", path)
}

File: clients/ts/conventions/goals/test/conformance.test.ts
// The TS goals conformance replay: it runs the REAL TS goal verbs against the TS
// Recorder and asserts the captured operations equal the language-neutral vectors
// under protocol/conformance/vectors/goals — the SAME JSON files the Go suite
// replays (FORMAT.md, ADR-0041). Passing the IDENTICAL setCriterion.json vector
// the Go suite passes is the op-transcript co-equality proof (AC#2).
//
// The comparison reproduces FORMAT.md exactly: order-sensitive ops; per-op op /
// subject / name / expectedRev equality; each payload compared as canonical JSON
// via the SDK's `canonical` (the single rule both languages must agree on).

import { test } from "node:test";
import assert from "node:assert/strict";
import { readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { canonical, type JSONValue } from "@sextant/sdk";
import {
  setCriterion,
  artifactName,
  StatusInProgress,
  StatusNotStarted,
  type SetCriterionInput,
} from "../src/index.js";
import { Recorder, type Op } from "./recorder.js";
import { goalVectorsDir } from "./repoRoot.js";

// fixedNow is the timestamp the recorded setCriterion verb stamps, so the captured
// goal.update is byte-stable and matches the vector's `updated` field. The peer of
// Go's fixedNow in conformance_test.go. The live SetCriterion takes the real time;
// only the recorded/replayed verb pins it.
const fixedNow = "2026-06-19T00:00:00Z";

// OpTranscriptVector is the on-disk vector shape (protocol/conformance/vector.go's
// OpTranscriptVector / FORMAT.md). input is opaque to the format; the verb decodes
// it.
interface OpTranscriptVector {
  epoch: number;
  convention: string;
  verb: string;
  description?: string;
  input: JSONValue;
  operations: Op[];
}

function loadGoalVectors(): Array<{ path: string; vector: OpTranscriptVector }> {
  const dir = goalVectorsDir();
  const files = readdirSync(dir).filter((f) => f.endsWith(".json"));
  if (files.length === 0) {
    throw new Error(`no goals vectors found under ${dir}`);
  }
  return files
    .sort()
    .map((f) => {
      const path = join(dir, f);
      return { path, vector: JSON.parse(readFileSync(path, "utf8")) as OpTranscriptVector };
    });
}

// runVerb dispatches (convention, verb) to the registered TS verb, run against the
// recorder. Only setCriterion exists today; an unknown verb fails loudly (a vector
// no verb answers is a gap, not a pass — mirroring the Go runner).
async function runVerb(rec: Recorder, v: OpTranscriptVector): Promise<void> {
  if (v.convention !== "goals") {
    throw new Error(`vector names convention ${JSON.stringify(v.convention)}, not goals`);
  }
  switch (v.verb) {
    case "setCriterion": {
      const input = v.input as unknown as SetCriterionInput;
      await setCriterion(rec, input, fixedNow);
      return;
    }
    default:
      throw new Error(`vector names verb ${JSON.stringify(v.verb)}, which is not registered`);
  }
}

// seedGoal seeds the recorder with the prior goal artifact a setCriterion vector
// reads before it writes (a read-then-write verb). The seed mirrors the bus state
// it would find live and does NOT appear in the transcript. It is the exact peer of
// the Go suite's seedGoal: a goal with the target criterion (starting not-started,
// owner sirius) plus a sibling (in-progress), at revision 4 — so the transcript
// shows the rewrite preserves siblings and the north-star and CAS's against rev 4.
function seedGoal(rec: Recorder, v: OpTranscriptVector): void {
  if (v.verb !== "setCriterion") return;
  const input = v.input as unknown as SetCriterionInput;
  const goal: JSONValue = {
    northstar: "Ship the goals convention",
    criteria: [
      { id: input.criterionId, text: "the criterion under test", status: StatusNotStarted, owner: "sirius" },
      { id: "other", text: "a sibling criterion", status: StatusInProgress },
    ],
  };
  rec.seedArtifact(artifactName(input.goalId), goal, 4);
}

// canonicalPayload canonicalizes an op's payload under the FORMAT.md rule. An
// absent payload canonicalizes to "null" (so two ops that both omit a payload
// compare equal) — mirroring Go's Canonicalize(nil) == "null".
function canonicalPayload(p: JSONValue | undefined): string {
  return canonical(p === undefined ? null : p);
}

// assertOpsEqual compares the recorded transcript to the vector's, order-sensitive,
// each payload under the canonical-JSON rule — the TS peer of the Go runner's
// assertOpsEqual.
function assertOpsEqual(path: string, want: Op[], got: Op[]): void {
  assert.equal(
    got.length,
    want.length,
    `${path}: operation count\n want: ${JSON.stringify(want.map((o) => o.op))}\n got:  ${JSON.stringify(got.map((o) => o.op))}`,
  );
  for (let i = 0; i < want.length; i++) {
    const w = want[i]!;
    const g = got[i]!;
    assert.equal(g.op, w.op, `${path}: op[${i}].op`);
    assert.equal(g.subject ?? "", w.subject ?? "", `${path}: op[${i}] (${w.op}).subject`);
    assert.equal(g.name ?? "", w.name ?? "", `${path}: op[${i}] (${w.op}).name`);
    assert.equal(
      g.expectedRev ?? null,
      w.expectedRev ?? null,
      `${path}: op[${i}] (${w.op}).expectedRev`,
    );
    const cw = canonicalPayload(w.payload);
    const cg = canonicalPayload(g.payload);
    assert.equal(cg, cw, `${path}: op[${i}] (${w.op}).payload mismatch\n want: ${cw}\n got:  ${cg}`);
  }
}

const vectors = loadGoalVectors();

test("the goals conformance vectors are discovered", () => {
  assert.ok(vectors.length >= 1, "expected at least one goals vector");
  // The setCriterion.json vector the Go suite passes must be present — the
  // identical file is the co-equality contract.
  assert.ok(
    vectors.some((v) => v.path.endsWith("setCriterion.json")),
    "expected the setCriterion.json vector the Go suite passes",
  );
});

for (const { path, vector } of vectors) {
  const rel = path.split("/").slice(-2).join("/");
  test(`goals vector ${rel} replays to the identical operations`, async () => {
    assert.equal(vector.epoch, 1, "the shipped goals vector is pinned to epoch 1");
    const rec = new Recorder();
    seedGoal(rec, vector);
    await runVerb(rec, vector);
    assertOpsEqual(path, vector.operations, rec.operations());
  });
}
